### PR TITLE
feat: tss-history

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description
- Add `add Tss History to crosschain store`
- Add `tss pubkey to createCCTX`
- Add a new admin function to update TSS manually instead of doing it automatically when a new TSS is generated
Closes: #1026 #1027

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: cmd/zetaclientd/keygen_tss.go [hunk 1]
```diff
@@ -16,9 +17,9 @@ import (
 	"gitlab.com/thorchain/tss/go-tss/p2p"
 )
 
-func GenerateTss(logger zerolog.Logger, cfg *config.Config, zetaBridge *mc.ZetaCoreBridge, peers p2p.AddrList, priKey secp256k1.PrivKey, ts *mc.TelemetryServer) (*mc.TSS, error) {
+func GenerateTss(logger zerolog.Logger, cfg *config.Config, zetaBridge *mc.ZetaCoreBridge, peers p2p.AddrList, priKey secp256k1.PrivKey, ts *mc.TelemetryServer, tssList []*types.TSS) (*mc.TSS, error) {
```
Comment (lumtis) [inline cmd/zetaclientd/keygen_tss.go:20]:
```
tssList []*types.TSS
```
Is it the TSS history? I would rename it `tssHistoricalList` to be more explicit

Comment (lumtis) [inline cmd/zetaclientd/keygen_tss.go:20]: Do we need using array of pointer? It seems to me that we never use `nil` value?

File: cmd/zetaclientd/start.go [hunk 2]
```diff
@@ -117,7 +117,12 @@ func start(_ *cobra.Command, _ []string) error {
 		}
 	}()
 
-	tss, err := GenerateTss(masterLogger, cfg, zetaBridge, peers, priKey, telemetryServer)
+	tssHistoricalList, err := zetaBridge.GetTssHistory()
+	if err != nil {
+		startLogger.Error().Err(err).Msg("GetTssHistory error")
+	}
+
+	tss, err := GenerateTss(masterLogger, cfg, zetaBridge, peers, priKey, telemetryServer, tssHistoricalList)
```
Comment (lumtis) [inline cmd/zetaclientd/start.go:125]: Maybe we should explicitly set an empty array for `tssHistoricalList` if the previous method returned an error

File: common/pubkey.go [hunk 3]
```diff
@@ -31,6 +32,18 @@ var EmptyPubKey PubKey
 
 // NewPubKey create a new instance of PubKey
 // key is bech32 encoded string
```
Comment (lumtis) [inline common/pubkey.go:34]: Misplaced comments

File: common/pubkey.go [hunk 4]
```diff
@@ -31,6 +32,18 @@ var EmptyPubKey PubKey
...
+
+func GetAddressFromPubkeyString(pubkey string) (sdk.AccAddress, error) {
```
Comment (lumtis) [inline common/pubkey.go:36]: Maybe we could add unit test for this method

File: proto/crosschain/genesis.proto [hunk 5]
```diff
@@ -24,4 +24,5 @@ message GenesisState {
 
   repeated LastBlockHeight lastBlockHeightList = 8;
   repeated InTxHashToCctx inTxHashToCctxList = 9 [(gogoproto.nullable) = false];
+  repeated TSS tss_history = 10;
```
Comment (lumtis) [inline proto/crosschain/genesis.proto:27]: Any reason to not use `[(gogoproto.nullable) = false]`? I feel like using pointers increase complexity and the risk of getting nil dereference

File: x/crosschain/client/cli/cli_tss.go [hunk 6]
```diff
@@ -41,6 +41,32 @@ func CmdShowTSS() *cobra.Command {
 	return cmd
 }
 
+func CmdListTssHistory() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list-tss-history",
```
Comment (lumtis) [inline x/crosschain/client/cli/cli_tss.go:46]: Maybe we could rename the query cmd into `TssHistory` instead of `ListTssHistory`

File: x/crosschain/genesis.go [hunk 7]
```diff
@@ -42,6 +42,10 @@ func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState)
 		k.SetTSS(ctx, *genState.Tss)
 	}
 
+	for _, elem := range genState.TssHistory {
+		k.SetTSSHistory(ctx, *elem)
```
Comment (lumtis) [inline x/crosschain/genesis.go:46]:
```suggestion

	for _, elem := range genState.TssHistory {
		if elem != nil {
		    k.SetTSSHistory(ctx, *elem)
		}
```

File: x/crosschain/genesis.go [hunk 8]
```diff
@@ -84,6 +88,11 @@ func ExportGenesis(ctx sdk.Context, k keeper.Keeper) *types.GenesisState {
 		genesis.CrossChainTxs = append(genesis.CrossChainTxs, &e)
 	}
 
+	tssList := k.GetAllTSS(ctx)
+	for _, elem := range tssList {
+		e := elem
+		genesis.TssHistory = append(genesis.TssHistory, e)
+	}
```
Comment (lumtis) [inline x/crosschain/genesis.go:95]: Can't we do `genesis.TssHistory = k.GetAllTSS(ctx)` directly?

File: x/crosschain/keeper/keeper_cross_chain_tx_vote_inbound_tx.go [hunk 9]
```diff
@@ -70,6 +70,10 @@ func (k msgServer) VoteOnObservedInboundTx(goCtx context.Context, msg *types.Msg
 	if receiverChain == nil {
 		return nil, sdkerrors.Wrap(types.ErrUnsupportedChain, fmt.Sprintf("ChainID %d, Observation %s", msg.ReceiverChain, observationType.String()))
 	}
+	tss, found := k.GetTSS(ctx)
+	if !found {
+		ctx.Logger().Error("TSS not found")
```
Comment (lumtis) [inline x/crosschain/keeper/keeper_cross_chain_tx_vote_inbound_tx.go:75]: Shouldn't we return an error? We use `tss.TssPubkey` later which I believe would cause a panic

File: x/crosschain/keeper/keeper_tss.go [hunk 10]
```diff
@@ -36,22 +45,32 @@ func (k Keeper) RemoveTSS(ctx sdk.Context) {
...
+		list = append(list, &val)
 	}
-
 	return
 }
 
+// GetPreviousTSS returns the previous tss information
+func (k Keeper) GetPreviousTSS(ctx sdk.Context) (val types.TSS, found bool) {
+	tssList := k.GetAllTSS(ctx)
+	if len(tssList) <= 2 {
```
Comment (lumtis) [inline x/crosschain/keeper/keeper_tss.go:64]:
```suggestion
	if len(tssList) < 2 {
```
I believe you can have a previous tss if len==2

File: x/crosschain/keeper/keeper_tss_voter.go [hunk 11]
```diff
@@ -89,13 +89,15 @@ func (k msgServer) CreateTSSVoter(goCtx context.Context, msg *types.MsgCreateTSS
...
+		tss := types.TSS{
 			TssPubkey:           msg.TssPubkey,
 			TssParticipantList:  keygen.GetGranteePubkeys(),
 			OperatorAddressList: ballot.VoterList,
 			FinalizedZetaHeight: ctx.BlockHeight(),
 			KeyGenZetaHeight:    msg.KeyGenZetaHeight,
-		})
+		}
+		k.SetTSS(ctx, tss)
+		k.SetTSSHistory(ctx, tss)
```
Comment (lumtis) [inline x/crosschain/keeper/keeper_tss_voter.go:100]: Maybe we could implement a method `AppendTSS` that would make the atomic operation of adding a new TSS.

File: x/crosschain/keeper/msg_server_whitelist_erc20.go [hunk 12]
```diff
@@ -35,6 +36,10 @@ func (k Keeper) WhitelistERC20(goCtx context.Context, msg *types.MsgWhitelistERC
 			return nil, sdkerrors.Wrapf(types.ErrInvalidAddress, "ERC20 contract address (%s) already whitelisted on chain (%d)", msg.Erc20Address, msg.ChainId)
 		}
 	}
+	tss, found := k.GetTSS(ctx)
+	if !found {
+		return nil, errorsmod.Wrapf(types.ErrCannotFindTSSKeys, "Cannot create new admin cmd")
```
Comment (lumtis) [inline x/crosschain/keeper/msg_server_whitelist_erc20.go:41]: > `Cannot create new admin cmd`

Is it the right error message? It seems not related to the error to me

File: zetaclient/tss_signer.go [hunk 13]
```diff
@@ -334,23 +339,50 @@ func getKeyAddrBTCWitnessPubkeyHash(tssPubkey string) (*btcutil.AddressWitnessPu
...
+			return fmt.Errorf("pubkey %s not found in keyshare", t.TssPubkey)
+		}
+	}
+	return nil
+}
+
+func (tss *TSS) LoadTssFilesFromDirectory(tssPath string) error {
+	files, err := os.ReadDir(tssPath)
+	if err != nil {
+		fmt.Println("ReadDir error", err)
```
Comment (lumtis) [inline zetaclient/tss_signer.go:383]:
```suggestion
		fmt.Println("ReadDir error:", err.Error())
```

File: cmd/zetaclientd/keygen_tss.go (hunk 1 above)
Comment (kingpinXD) [reply cmd/zetaclientd/keygen_tss.go:20]: We don't need it , but I don't see any issues with keeping it. Do you see any issues with the pointer ?

File: common/pubkey.go (hunk 4 above)
Comment (kingpinXD) [reply common/pubkey.go:36]: This method is not needed, I removed it.

File: x/crosschain/client/cli/cli_tss.go (hunk 6 above)
Comment (kingpinXD) [reply x/crosschain/client/cli/cli_tss.go:46]: We have been using `show-XXX` and `list-XXX` for all items to show query single and multiple . 
This goes with that nomenclature

File: x/crosschain/keeper/keeper_cross_chain_tx_vote_inbound_tx.go (hunk 9 above)
Comment (kingpinXD) [reply x/crosschain/keeper/keeper_cross_chain_tx_vote_inbound_tx.go:75]: We shouldn't return here, as deposit Tx does not need a TSS address. I added some conditional logic instead. Thanks for pointing this out

File: x/crosschain/keeper/msg_server_whitelist_erc20.go (hunk 12 above)
Comment (kingpinXD) [reply x/crosschain/keeper/msg_server_whitelist_erc20.go:41]: whitelistERC20 is an action of type admin cmd, we will have more of these in the future .
Added some details to the error print so that it makes more sense.

File: zetaclient/tss_signer.go [hunk 14]
```diff
@@ -334,23 +339,50 @@ func getKeyAddrBTCWitnessPubkeyHash(tssPubkey string) (*btcutil.AddressWitnessPu
...
+	}
+	err = tss.VerifyKeysharesForPubkeys(tssHistoricalList)
 	if err != nil {
-		fmt.Println("ReadDir error", err)
 		return nil, err
 	}
+	return &tss, nil
+}
+
+func (tss *TSS) VerifyKeysharesForPubkeys(tssList []types.TSS) error {
```
Comment (brewmaster012) [inline zetaclient/tss_signer.go:366]: i don't think this is correct, new keysigners would not have old keyshares that they were not part of.

**Review by brewmaster012 — CHANGES_REQUESTED** [review]

File: zetaclient/tss_signer.go (hunk 14 above)
Comment (kingpinXD) [reply zetaclient/tss_signer.go]: Thanks that's a great catch , 
Modified to using only TSS that the node participated in

**Comment by lumtis** [comment]
Is it ready for re-review @kingpinXD ?

**Review by brewmaster012 — APPROVED** [review]

File: x/observer/keeper/observer_mapper.go [hunk 15]
```diff
@@ -93,24 +93,32 @@ func (k msgServer) AddObserver(goCtx context.Context, msg *types.MsgAddObserver)
 	if msg.Creator != k.GetParams(ctx).GetAdminPolicyAccount(types.Policy_Type_add_observer) {
 		return &types.MsgAddObserverResponse{}, types.ErrNotAuthorizedPolicy
 	}
+	pubkey, _ := common.NewPubKey(msg.ZetaclientGranteePubkey)
+	granteeAddress, _ := common.GetAddressFromPubkeyString(msg.ZetaclientGranteePubkey)
+	k.DisableInboundOnly(ctx)
+	// AddNodeAccountOnly flag usage
+	// True: adds observer into the Node Account list but returns without adding to the observer list
+	// False: adds observer to the observer list, and not the node account list
```
Comment (lumtis) [inline x/observer/keeper/observer_mapper.go:101]: The `only` naming is a bit confusing to me because it would imply it's an action that is always executed but only this action is executed if the flag is set.

File: zetaclient/bitcoin_client.go [hunk 16]
```diff
@@ -183,6 +183,28 @@ func (ob *BitcoinChainClient) GetLastBlockHeight() int64 {
 	return height
 }
 
+func (ob *BitcoinChainClient) SetLastBlockHeightScanned(block int64) {
```
Comment (lumtis) [inline zetaclient/bitcoin_client.go:186]: I think we should add comment to explain difference with `GetLastBlockHeight`

File: zetaclient/chainclient.go [hunk 17]
```diff
@@ -10,8 +10,8 @@ import (
 // general chain client
 
 type ChainClient interface {
-	GetLastBlockHeight() int64 // 0 means error
-	SetLastBlockHeight(int64)
+	//GetLastBlockHeight() int64 // 0 means error
+	//SetLastBlockHeight(int64)
```
Comment (lumtis) [inline zetaclient/chainclient.go:14]: I think we should remove methods in interfaces no longer used instead of commenting them out if we plan to no longer use them. It might be confusing when reading the code.

File: zetaclient/tss_signer.go [hunk 18]
```diff
@@ -359,23 +364,60 @@ func getKeyAddrBTCWitnessPubkeyHash(tssPubkey string) (*btcutil.AddressWitnessPu
...
+		if NodeWasPartOfTss(granteePubKey32, t.TssParticipantList) {
+			if _, ok := tss.Keys[t.TssPubkey]; !ok {
+				return fmt.Errorf("pubkey %s not found in keyshare", t.TssPubkey)
+			}
+		}
+	}
+	return nil
+}
+
+func NodeWasPartOfTss(granteePubKey32 string, granteeList []string) bool {
```
Comment (lumtis) [inline zetaclient/tss_signer.go:407]:
```suggestion
func WasNodePartOfTss(granteePubKey32 string, granteeList []string) bool {
```

File: x/observer/client/cli/tx_add_observer.go [hunk 19]
```diff
@@ -10,7 +13,7 @@ import (
 
 func CmdAddObserver() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "add-observer [observer-address] [zetaclient-grantee-pubkey]",
+		Use:   "add-observer [observer-address] [zetaclient-grantee-pubkey] [add_node_account_only]",
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/observer/client/cli/tx_add_observer.go:16]: I agree, but this flag is temporary only, the only reason we are doing this in two steps as because we don't have the post-generation tasks ready, and we need to push the upgrade out soon. This flag will probably be removed in the next release.

**Review by kevinssgh — APPROVED** [review]
Looks good
